sklearn_numba_dpex/kmeans/kernels/utils.py

```diff
@@ -25,6 +25,150 @@
 import numba_dpex as dpex
 
 
+@lru_cache
+def make_relocate_empty_clusters_kernel(
+    n_relocated_clusters,
+    n_features,
+    n_selected_gt_threshold,
+    work_group_size,
+    dtype,
+):
+    n_work_groups_for_cluster = math.ceil(n_features / work_group_size)
+    n_work_items_for_cluster = n_work_groups_for_cluster * work_group_size
+    global_size = n_work_items_for_cluster * n_relocated_clusters
+
+    n_selected_gt_threshold = n_selected_gt_threshold - 1
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    # fmt: off
+    def relocate_empty_clusters(
+        X_t,                        # IN READ-ONLY   (n_features, n_samples)
+        sample_weight,              # IN READ-ONLY   (n_samples,)
+        assignments_idx,            # IN             (n_samples,)
+        samples_far_from_center,    # IN             (n_samples,)
+        empty_clusters_list,        # IN             (n_clusters,)
+        per_sample_inertia,         # INOUT          (n_samples,)
+        centroids_t,                # INOUT          (n_features, n_clusters)
+        cluster_sizes               # INOUT          (n_clusters,)
+            ):
+    # fmt: on
+        """NB: because of how the array was created, samples_far_from_center values
+        are located between (n_selected_gt_threshold - relocated_idx) and
+        (relocated_idx) indices.
+        """
+        group_idx = dpex.get_group_id(0)
+        item_idx = dpex.get_local_id(0)
+        relocated_idx = group_idx // n_work_groups_for_cluster
+        feature_idx = ((group_idx % n_work_groups_for_cluster) * work_group_size) + item_idx
+
+        if feature_idx >= n_features:
+            return
+
+        relocated_cluster_idx = empty_clusters_list[relocated_idx]
+        new_location_X_idx = samples_far_from_center[n_selected_gt_threshold - relocated_idx]
+        new_location_previous_assignment = assignments_idx[new_location_X_idx]
+
+        new_centroid_value = X_t[feature_idx, new_location_X_idx]
+        new_location_weight = sample_weight[new_location_X_idx]
+        X_centroid_addend = new_centroid_value * new_location_weight
+
+        # Cancel the contribution to the updated centroids of the sample that was once
+        # assigned to new_location_previous_assignment but is now assigned to the
+        # cluster of the centroids that relocates to this sample
+        dpex.atomic.sub(
+            centroids_t,
+            (feature_idx, new_location_previous_assignment),
+            X_centroid_addend
+            )
+
+        # The relocated centroid has only one contribution now, which is the sample
+        # to which it has been relocated to
+        centroids_t[feature_idx, relocated_cluster_idx] = X_centroid_addend
+
+        # Likewise, we update the weights in the clusters
+        if feature_idx == zero:
```

Comment:
`feature_idx` is likely an integral whilst `zero` is likely a floating. I do not think this causes problems, but I am just checking.

Reply:
I'm not surprised that it compiles, it seems like in numba_dpex built in functions have as much implementations as there are possible combinations of input types, if an implementation is not available for a given combination of types the compilation fail with an explicit error message that lists all possible implementations and explains what is the combination of types was rejected. I'm more worried of implicit casting within those implementations.